File: c/sum-of-multiples/sum_of_multiples.c
#include "sum_of_multiples.h"

unsigned int sum(const unsigned int* factors, const size_t number_of_factors, const unsigned int limit) {
    unsigned int ans = 0;
    for (unsigned int i = 1; i < limit; i++) {
        for (size_t j = 0; j < number_of_factors; j++) {
            if (factors[j] != 0 && i % factors[j] == 0) {
                ans += i;
                break;
            }
        }
    }
    return ans;
}

File: zig/sum-of-multiples/sum_of_multiples.zig
const std = @import("std");
const mem = std.mem;

pub fn sum(allocator: mem.Allocator, factors: []const u32, limit: u32) !u64 {
    var multiples = std.hash_map.AutoHashMap(u64, void).init(allocator);
    defer multiples.deinit();

    for (factors) |factor| {
        if (factor == 0) continue;
        var num = factor;
        while (num < limit) : (num += factor) {
            try multiples.put(num, {});
        }
    }

    var ans: u64 = 0;
    var keys = multiples.keyIterator();
    while (keys.next()) |key| ans += key.*;
    return ans;
}
